### PR TITLE
make all_equal() faster

### PR DESCRIPTION
**Pull request by fyrchik**
Hello!
This PR adresses #282 issue. Variant with `dedup` does not short circuit, but one with `all` does.
I have also added some benchmarks and test for an empty iterator.
```
test all_equal                                ... bench:     999,832 ns/iter (+/- 217,245)
test all_equal_default                        ... bench:   4,814,277 ns/iter (+/- 315,335)
test all_equal_for                            ... bench:   2,096,174 ns/iter (+/- 165,596)
```

Let me know, what do you think.

### CODE REVIEW

**Comment by timvermeulen** [comment]
Looks good!

**Comment by axelf4** [comment]
An alternative implementation could be:

```rust
self.try_fold(None, |acc, x| acc.filter(|&prev| x != prev).xor(Some(Some(x))))
```

**Comment by timvermeulen** [comment]
@axelf4 I haven't done any benchmarks, have you by any chance? I'd be worried that that would perform worse than this PR's implementation because of the extra branching in the closure body. But it's very hard to tell without benchmarks.

File: src/lib.rs [hunk 1]
```diff
@@ -1310,9 +1310,13 @@ pub trait Itertools : Iterator {
     /// assert!(data.into_iter().all_equal());
     /// ```
     fn all_equal(&mut self) -> bool
-        where Self::Item: PartialEq,
+        where Self: Sized,
```
Comment (phimuemue) [inline src/lib.rs]: I think requiring `Sized` is ok. Can anyone with more expertise confirm this is unproblematic?

File: tests/test_std.rs [hunk 2]
```diff
@@ -100,6 +100,7 @@ fn dedup() {
 
 #[test]
 fn all_equal() {
+    assert!("".chars().all_equal());
```
Comment (phimuemue) [inline tests/test_std.rs]: Good one.

File: src/lib.rs (hunk 1 above)
Comment (timvermeulen) [reply src/lib.rs]: There are already lots of methods you can't call on an unsized iterator, so I don't think this is problematic.

Comment (timvermeulen) [reply src/lib.rs]: I believe the documentation (sometimes?) leaves out `Self: Sized` bounds because they’re so common 😕

Comment (timvermeulen) [reply src/lib.rs]: I believe the documentation (sometimes?) leaves out `Self: Sized` bounds because they’re so common 😕

Comment (jswrenn) [reply src/lib.rs]: Per [Reddit](https://www.reddit.com/r/rust/comments/cgh56g/hey_rustaceans_got_an_easy_question_ask_here/eukw73y/?context=1&depth=3), I think this is a bug. I've filed an issue: https://github.com/rust-lang/rust/issues/62899.

File: tests/test_std.rs (hunk 2 above)
Comment (fyrchik) [reply tests/test_std.rs]: Yes, it will be a nice addtition. I have added it.

Comment (bluss) [reply tests/test_std.rs]: Adding a quickcheck test would be a great alternative to either of those (something for another PR)

**Comment by bluss** [comment]
This should not need to require Self: Sized, and I'd prefer to rewrite it so that it is not a breaking change (in my opinion).

I'm not sure I understand the short circuit argument - in what sense does this short circuit more than dedup does? They should both visit the same number of iterator elements.

**Comment by timvermeulen** [comment]
@bluss The `dedup` version does also short-circuit, but only after the second "group" has been iterated entirely, rather than right after the first element of the second group.

**Comment by bluss** [comment]
Oh. We are long overdue for this improvement then.

One major part of benchmark speed up could be that `all` is explicitly unrolled in the slice iterator (slice iterator specific, in its `try_fold`). That's nice to use, but we can note that it is not a general improvement. In addition to that we have the improvement of using internal iteration (all/try_fold etc) which is a bit more general.

Looking at those upsides it seems ok to have a breaking change. I'm not sure in what context it would be a detectable breaking change, it should only change which implementation is picked when we call this method on a trait object.

... and it turns out the trait object question is moot, because trait object support has been broken in Itertools 0.8.0, `dyn Itertools` is not possible at the moment precisely because of some missing Self: Sized bounds.

That's something we could revisit, but let's only add back trait object support if we can find a good reason. Clearly there was no test for it.

With that, I don't think adding `Self: Sized` is a breaking change.

**Comment by jswrenn** [comment]
If the trait object question is moot, I'm happy to merge this!

bors r+
